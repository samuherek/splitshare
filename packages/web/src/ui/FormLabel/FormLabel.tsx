import clsx from 'clsx';
import React from 'react';
import styled from 'styled-components';
import { useFormControlCtx } from '../FormControl/FormControlProvider';
import getFormControlState from '../FormControl/getFormControlState';

interface Props {
  children: React.ReactNode;
  className?: string;
  component?: any;
  disabled?: boolean;
  error?: boolean;
  filled?: boolean;
  focused?: boolean;
  required?: boolean;
}

export const classes = {
  root: 'SSFormLabel',
  optional: 'SSFromLabelOptional',
  disabled: 'disabled',
  focused: 'focused',
  error: 'error',
  required: 'required',
  filled: 'filled',
};

const LabelStyled = styled.label`
  line-height: 1;
  padding: 0;
  font-size: 1em;
  color: inherit;
`;

const FormLabel = React.forwardRef<Props, any>(function FormLabel(props, ref) {
  const {
    children,
    className,
    component: Component = LabelStyled,
    disabled,
    error,
    filled,
    focused,
    required,
    ...rest
  } = props;

  const ctxFormControl = useFormControlCtx();
  const fcs: any = getFormControlState({
    props,
    ctxFormControl,
    states: ['disabled', 'focused', 'error', 'required', 'filled'],
  });

  return (
    <Component
      className={clsx(className, classes.root, {
        [classes.disabled]: fcs.disabled,
        [classes.error]: fcs.error,
        [classes.filled]: fcs.filled,
        [classes.focused]: fcs.focused,
        [classes.required]: fcs.required,
      })}
      ref={ref}
      {...rest}
    >
      {children}
      {!fcs.required && (
        <span
          className={clsx(classes.optional, {
            [classes.error]: fcs.error,
          })}
        >
          &thinsp;{'(Optional)'}
        </span>
      )}
    </Component>
  );
});

export default FormLabel;
