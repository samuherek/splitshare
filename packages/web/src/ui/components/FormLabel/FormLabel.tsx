import React from 'react';
import styled from 'styled-components';

interface FormLabelProps {
  children: React.ReactNode;
  className?: string;
  component?: React.ReactNode;
  disabled?: boolean;
  error?: boolean;
  filled?: boolean;
  focused?: boolean;
  required?: boolean;
}

const LabelStyled = styled.label`
  line-height: 1;
  padding: 0;
  font-size: 1em;
  color: rgba(0, 0, 0, 0.54);
`;

// TODO: Add the FormControl context into this business ;)
const FormLabel = React.forwardRef<any, FormLabelProps>(function FormLabel(
  props,
  ref
) {
  const {
    children,
    className,
    component: Component = LabelStyled,
    disabled,
    error,
    filled,
    focused,
    required,
    ...rest
  } = props;

  return (
    // @ts-ignore
    <Component className={className} ref={ref} {...rest}>
      {children}
      {required && <span>&thinsp;{'*'}</span>}
    </Component>
  );
});

export default FormLabel;
