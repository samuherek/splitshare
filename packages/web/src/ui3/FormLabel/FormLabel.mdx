---
name: FormLabel
---

import { Playground, Props } from 'docz';
import FormLabel from './FormLabel';

# FormLabel

Form label component is to be used for any labeing in a "Form". It's to add labels for checkbox or radio groups.

### Props

<Props of={FormLabel} />

### Playground

<Playground>
  <FormLabel />
</Playground>
