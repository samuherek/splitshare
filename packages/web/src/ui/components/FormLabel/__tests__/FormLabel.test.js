import React from 'react';
import { render } from '@testing-library/react';
import FormLabel from '../FormLabel';

describe('<FormLabel />', () => {
  it('should render snapshot', () => {
    const { getByText } = render(<FormLabel>Label</FormLabel>);
    expect(getByText(/label/i)).toMatchSnapshot();
  });
});
