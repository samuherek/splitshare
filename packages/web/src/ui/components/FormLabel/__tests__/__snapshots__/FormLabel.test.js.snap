// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormLabel /> should render snapshot 1`] = `
.c0 {
  line-height: 1;
  padding: 0;
  font-size: 1em;
  color: rgba(0,0,0,0.54);
}

<label
  class="c0"
>
  Label
</label>
`;
